File: src/main/java/page/battle/RecdSavePage.java
package page.battle;

import common.util.stage.Replay;
import io.BCMusic;
import main.MainBCU;
import main.Opts;
import page.JBTN;
import page.JTF;
import page.Page;

import javax.swing.*;
import java.awt.event.FocusAdapter;
import java.awt.event.FocusEvent;

public class RecdSavePage extends Page {

	private static final long serialVersionUID = 1L;

	private final JBTN back = new JBTN(0, "back");
	private final JBTN save = new JBTN(0, "save");
	private final JTF jtf = new JTF();

	private final Replay recd;
	private String name;

	protected RecdSavePage(Page p, Replay rec) {
		super(p);
		recd = rec;

		ini();
	}

	@Override
	protected JButton getBackButton() {
		return back;
	}

	@Override
	protected void resized(int x, int y) {
		setBounds(0, 0, x, y);

		set(back, x, y, 0, 0, 200, 50);
		set(jtf, x, y, 1000, 500, 300, 50);
		set(save, x, y, 1000, 600, 300, 50);
	}

	private void addListeners() {
		back.addActionListener(arg0 -> changePanel(getFront()));

		jtf.addFocusListener(new FocusAdapter() {
			@Override
			public void focusLost(FocusEvent arg0) {
				String str = jtf.getText().trim();
				if (str.length() == 0)
					str = "new replay " + recd.st.toString();
				str = MainBCU.validate(str, '#');
				jtf.setText(name = str);
			}
		});

		save.addActionListener(arg0 -> {
			if (Replay.getMap().containsKey(name) && !Opts.conf("A replay named " + name + " already exists. Do you wish to overwrite?"))
				return;

			recd.rename(name, true);
			recd.write();
			Replay.getMap().put(recd.rl.id, recd);

			if (BCMusic.music != null)
				BCMusic.stopAll();

			changePanel(new RecdManagePage(getRootPage()));
		});

	}

	private void ini() {
		add(back);
		add(jtf);
		add(save);
		addListeners();
		String initName = name = "new replay " + recd.st.toString().replaceAll("/", "-");
		jtf.setText(initName);
	}

}

File: src/page/battle/RecdSavePage.kt
package page.battle

import common.util.stage.Recd
import main.MainBCU
import page.JBTN
import page.JTF
import page.Page
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.FocusAdapter
import java.awt.event.FocusEvent

class RecdSavePage(p: Page?, rec: Recd) : Page(p) {
    private val back: JBTN = JBTN(0, "back")
    private val save: JBTN = JBTN(0, "save")
    private val jtf: JTF = JTF()
    private val recd: Recd
    override fun resized(x: Int, y: Int) {
        setBounds(0, 0, x, y)
        Page.Companion.set(back, x, y, 0, 0, 200, 50)
        Page.Companion.set(jtf, x, y, 1000, 500, 300, 50)
        Page.Companion.set(save, x, y, 1000, 600, 300, 50)
    }

    private fun addListeners() {
        back.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                changePanel(front)
            }
        })
        jtf.addFocusListener(object : FocusAdapter() {
            override fun focusLost(arg0: FocusEvent?) {
                var str: String = jtf.getText().trim { it <= ' ' }
                str = MainBCU.validate(str)
                if (str.length == 0) str = "new record"
                str = Recd.Companion.getAvailable(str)
                recd.name = str
            }
        })
        save.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                recd.write()
                Recd.Companion.map.put(recd.name, recd)
                changePanel(RecdManagePage(rootPage))
            }
        })
    }

    private fun ini() {
        add(back)
        add(jtf)
        add(save)
        jtf.setText(Recd.Companion.getAvailable("new record"))
        addListeners()
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        recd = rec
        ini()
        resized()
    }
}
